File: tools/markdown-viewer/watch-manifest.js
#!/usr/bin/env node
/**
 * File watcher: auto-run generate-manifest.py when .md files change.
 * Run: npm run watch  (requires: npm install chokidar)
 */

const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

let chokidar;
try {
  chokidar = require('chokidar');
} catch (e) {
  console.error('Install chokidar: npm install chokidar');
  process.exit(1);
}

const root = path.join(__dirname);
const watchPath = path.join(root, '..', 'MarkdownSV');
const manifestScript = path.join(root, 'generate-manifest.py');

function runManifest() {
  const env = { ...process.env, CONTENT_ROOT: path.resolve(root, '..', 'MarkdownSV') };
  const proc = spawn('python3', [manifestScript], { cwd: root, env, stdio: 'inherit' });
  proc.on('error', () => {
    console.error('Could not run generate-manifest.py');
  });
}

let debounce;
const watcher = chokidar.watch(watchPath, {
  ignored: /(^|[\/\\])(\.|node_modules)/,
  persistent: true
});

watcher.on('change', (p) => {
  if (!p.endsWith('.md')) return;
  clearTimeout(debounce);
  debounce = setTimeout(() => {
    console.log('\n📝 .md change detected → regenerating files.json');
    runManifest();
  }, 300);
});

watcher.on('add', (p) => {
  if (!p.endsWith('.md')) return;
  clearTimeout(debounce);
  debounce = setTimeout(() => {
    console.log('\n📝 New .md file → regenerating files.json');
    runManifest();
  }, 300);
});

watcher.on('unlink', (p) => {
  if (!p.endsWith('.md')) return;
  clearTimeout(debounce);
  debounce = setTimeout(() => {
    console.log('\n📝 .md removed → regenerating files.json');
    runManifest();
  }, 300);
});

watcher.on('addDir', () => {
  clearTimeout(debounce);
  debounce = setTimeout(() => {
    console.log('\n📁 New folder → regenerating files.json');
    runManifest();
  }, 300);
});

watcher.on('unlinkDir', () => {
  clearTimeout(debounce);
  debounce = setTimeout(() => {
    console.log('\n📁 Folder removed → regenerating files.json');
    runManifest();
  }, 300);
});

console.log('👀 Watching', watchPath, 'for .md and folder changes...');
console.log('   Add/edit/delete .md or new/removed folders → files.json auto-updates → refresh viewer');
console.log('   Press Ctrl+C to stop\n');

File: tool/MDviewer/watch-manifest.py
#!/usr/bin/env python3
"""
File watcher: auto-run generate-manifest.py when .md files change.
No commit needed — add/edit/delete .md → manifest updates → refresh viewer.
Requires: pip install watchdog (or use without: python3 watch-manifest.py --no-watch for one-shot)
"""

import os
import sys
import subprocess
import argparse
from pathlib import Path

try:
    from watchdog.observers import Observer
    from watchdog.events import FileSystemEventHandler
    HAS_WATCHDOG = True
except ImportError:
    HAS_WATCHDOG = False


def run_generate_manifest():
    """Run generate-manifest.py in current dir"""
    script = Path(__file__).parent / 'generate-manifest.py'
    subprocess.run([sys.executable, str(script)], cwd=Path(__file__).parent)


class ManifestHandler(FileSystemEventHandler):
    def __init__(self):
        super().__init__()
        self._debounce_sec = 0.5
        self._last_run = 0

    def on_any_event(self, event):
        if event.is_directory:
            return
        path = getattr(event, 'dest_path', None) or event.src_path
        if path and path.endswith('.md'):
            import time
            now = time.time()
            if now - self._last_run >= self._debounce_sec:
                self._last_run = now
                print("\n📝 .md change detected → regenerating files.json")
                run_generate_manifest()


def main():
    parser = argparse.ArgumentParser(description='Watch .md files and auto-update files.json')
    parser.add_argument('--no-watch', action='store_true', help='Run manifest once and exit (no watching)')
    args = parser.parse_args()

    root = Path(__file__).parent
    os.chdir(root)

    if args.no_watch or not HAS_WATCHDOG:
        if not HAS_WATCHDOG and not args.no_watch:
            print("⚠️  Install watchdog for live watching: pip install watchdog")
            print("   Running manifest once (use --no-watch to suppress this message)\n")
        run_generate_manifest()
        if args.no_watch or not HAS_WATCHDOG:
            return

    # Watch content/ if exists, else .
    watch_path = root / 'content' if (root / 'content').exists() else root
    event_handler = ManifestHandler()
    observer = Observer()
    observer.schedule(event_handler, str(watch_path), recursive=True)
    observer.start()

    print(f"👀 Watching {watch_path} for .md changes...")
    print("   Add/edit/delete .md → files.json auto-updates → refresh viewer")
    print("   Press Ctrl+C to stop\n")

    try:
        while True:
            import time
            time.sleep(1)
    except KeyboardInterrupt:
        observer.stop()
    observer.join()
    print("\n👋 Stopped.")


if __name__ == "__main__":
    main()
